test/common/dashboards_data.go

```diff
@@ -0,0 +1,352 @@
+		}
+	}
+
+	return pods, nil
+}
+
+func getRHMINamespaces(prometheusPodName string, ctx *TestingContext) ([]string, error) {
+	queryResult, err := queryPrometheus("kube_namespace_labels{label_monitoring_key='middleware'}", prometheusPodName, ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query prometheus: %w", err)
+	}
+
+	var namespaces []string
+
+	for _, result := range queryResult {
+		namespaces = append(namespaces, result.Metric["namespace"].(string))
+	}
+
+	return namespaces, nil
+}
+
+func getDashboardPanels(dashboardName string, grafanaPodName string, ctx *TestingContext) ([]panelDefinition, error) {
+	query := url.QueryEscape(dashboardName)
+	searchOutput, err := curlGrafana("/api/search?query="+query, grafanaPodName, ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to curl grafana: %w", err)
+	}
+
+	var dashboardSearch []dashboardSearchResponse
+	err = json.Unmarshal([]byte(searchOutput), &dashboardSearch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshall json: %w", err)
+	}
+
+	if len(dashboardSearch) != 1 {
+		return nil, fmt.Errorf(dashboardName + " dashboard not found")
+	}
+
+	dashboardOutput, err := curlGrafana("/api/dashboards/uid/"+dashboardSearch[0].UID, grafanaPodName, ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to curl grafana: %w", err)
+	}
+
+	var dashboardDetail dashboardDetailResponse
+	err = json.Unmarshal([]byte(dashboardOutput), &dashboardDetail)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshall json: %w", err)
+	}
+
+	return dashboardDetail.Dashboard.Panels, nil
+}
+
+func getMonitoringAppPodName(app string, ctx *TestingContext) (string, error) {
+	pods := &corev1.PodList{}
+	opts := []k8sclient.ListOption{
+		k8sclient.InNamespace(MonitoringOperatorNamespace),
+		k8sclient.MatchingLabels{"app": app},
+	}
+
+	err := ctx.Client.List(goctx.TODO(), pods, opts...)
+	if err != nil {
+		return "", fmt.Errorf("failed to list pods: %w", err)
+	}
+
+	if len(pods.Items) != 1 {
+		return "", fmt.Errorf("grafana pod not found")
+	}
+
+	return pods.Items[0].ObjectMeta.Name, nil
+}
+
+func curlGrafana(path string, podName string, ctx *TestingContext) (string, error) {
+	return execToPod("curl localhost:3000"+path,
+		podName,
+		MonitoringOperatorNamespace,
+		"grafana", ctx)
+}
+
+func queryPrometheus(query string, podName string, ctx *TestingContext) ([]prometheusQueryResult, error) {
+	queryOutput, err := execToPod("curl localhost:9090/api/v1/query?query="+url.QueryEscape(query),
+		podName,
+		MonitoringOperatorNamespace,
+		"prometheus", ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to exec to prometheus pod: %w", err)
+	}
+
+	return getPrometheusQueryResult(queryOutput)
+}
+
+func queryPrometheusMany(queries []string, podName string, ctx *TestingContext) ([]string, error) {
+	command := ""
+
+	for _, query := range queries {
+		command += "curl localhost:9090/api/v1/query?query=" + url.QueryEscape(query) + ";"
```

Comment:
Not a fan of this implementation, I'm afraid that if a single curl will fail for any reason we will not have any error message until we the result is Unmarshal and most probably it will fail at that point, even worst the single query could be completed ignored 

Have you tried to pass a wrong/broken query to see what happened?

We could probably do one execToPod for query if it is not excessive

Reply:
I tried to separate the curl calls but it will take at least 10 times more, so ignore it :sweat_smile: